Make two or more API versions work

davi5e opened the pull request:
So, I followed the README and turns out `api_route()` can't distinguish between 2 or more APIs...
The code only works for one of the many, more precisely the one defined in `config('api.default')`.

This is not a perfect solution to the problem of using arrays instead of strings in `config/api.php`, but at least now both APIs get their routes as intended...

For comparison, here is my api.php file:
```php
<?php

return [
    'default' => 'api',
    'version' => [
        'api' => '2',
        'desktop' => '3',
    ],
    'resources_path' => [
        'api' => 'App\Http\Resources',
        'desktop' => 'App\Http\Resources',
    ],
    'resources' => [
        'api' => 'Api',
        'desktop' => 'Desktop',
    ],
    'route_prefix' => [
        'api' => 'api',
        'desktop' => 'desktop',
    ],
];
```

And after applying the modifications to `getRouteName` both commands bellow work as expected:
```php
<?php

 # Interprets 'api.some.name' as 'route('api.v2.some.name')'
api_route('api.some.name');
 # Interprets 'desktop.another.name' as 'route('desktop.v3.another.name')'
api_route('desktop.another.name');
```

davi5e committed:
Make two or more API versions work

So, I followed the README and turns out api_route() can't distinguish between 2 or more APIs...
The code only works for one of the many, more precisely the one defined in `config('api.default')`.
This is not a perfect solution to the problem of using arrays instead of strings in `config/api.php`, but at least now both APIs get their routes as intended...

For comparison, here is my api.php file:
```php
<?php

return [
    'default' => 'api',
    'version' => [
        'api' => '2',
        'desktop' => '3',
    ],
    'resources_path' => [
        'api' => 'App\Http\Resources',
        'desktop' => 'App\Http\Resources',
    ],
    'resources' => [
        'api' => 'Api',
        'desktop' => 'DEsktop',
    ],
    'route_prefix' => [
        'api' => 'api',
        'desktop' => 'desktop',
    ],
];
```

And after applying the modifications to `getRouteName` both commands bellow work as expected:
```php
<?php
 # Interprets 'api.some.name' as 'route('api.v2.some.name')'
api_route('api.some.name');
 # Interprets 'desktop.another.name' as 'route('desktop.v3.another.name')'
api_route('desktop.another.name');
```

juampi92 commented:
Can you please add extra unit tests that contemplate that scenario that you are trying to fix then?
Cause your changes break the normal expected behavior (see the CI)

davi5e committed:
Keeping the original code as much as possible

In order to solve the unit test problem (I never coded one, sorry...) I changed the logic to preserve the original intent.

davi5e commented:
@juampi92 Did my best to keep the original code intact while also commenting every step to dynamically find the API version.

Don't know how to run the test (my bad, will improve), but should work now :)

juampi92 commented:
@davi5e , so I investigated you issue, and I was trying replicate it, but I encountered a problem:

What you are trying to do is, in the desktop route, you are trying to use a versioned api route. Or vice-versa.

The problem with this, is that you don't know the current of api, you only know the current of desktop.

Imagine having this config:
```php
return [
    'default' => 'api',
    'version' => [
        'api' => '2',
        'desktop' => '3',
    ],
    'resources_path' => 'App\Http\Resources',
    'resources' => [
        'api' => 'Api',
        'desktop' => 'Desktop',
    ],
    'route_prefix' => [
        'api' => 'api',
        'desktop' => 'desktop',
    ],
];
```

And you access route `/desktop/v1/posts`. If that response uses the api_route on api, which version should it use?
- `api/v2` (max)
- `api/v1` (like the current desktop)
- or `api/v3` (max of desktop, why not) ?

In your PR, you used the latest of `api`. It's a possible solution, but it's not necessarily correct on every situation.

Let me know what you think, but grabbing the latest version is not a propper solution.

davi5e commented:
In my mind, keeping the names we're using in this example, `api` and `desktop` are two completely separate APIs.

To further illustrate, they'd have different configurations under `app/Providers/RouteServiceProvider.php`, for example:
```php
<?php

    public function map()
    {
        $this->mapApiRoutes();
        $this->mapDesktopRoutes();
        $this->mapWebRoutes();
    }

    protected function mapApiRoutes()
    {
        // URL: https://example.com/api/v2/...
        Route::prefix('api')
             ->middleware('api')
             ->namespace('App\Http\Resources')
             ->group(base_path('routes/api.php'));
    }

    protected function mapDesktopRoutes()
    {
        // URL: https://example.com/desktop/v3/...
        Route::prefix('desktop')
             ->middleware('api')
             ->namespace('App\Http\Resources')
             ->group(base_path('routes/desktop.php'));
    }
```

Having a bunch of different APIs backed by the same Laravel backend can be very useful and this package would help to version them all, enabling a single call to `api_route()` to figure out every API version and return it's proper route.

This behavior does not exists today since it's only possible to expand upon one of the many APIs defined in `config/api.php`. The goal is to use `api_route()` to correctly yield any route of the APIs since the regex identifies its prefix and uses it as the arrays' key accordingly.

It's a simple expansion from `api_route('desktop.some.route')` to `route('desktop.v3.some.route')` but for any number of APIs. And the modifications do get the latest version of each since it uses the rest of the code as is...

Also, while I was making the last changes, I noticed this case of mine do not, in fact, require the default API to be used. Since the proposed changes always use the received prefix as a key in `config('api.resources')` there would be no need to a fallback option, as long as every call is consistent, like in my first post:
```php
<?php

 # Interprets 'api.some.name' as 'route('api.v2.some.name')'
api_route('api.some.name');
 # Interprets 'desktop.another.name' as 'route('desktop.v3.another.name')'
api_route('desktop.another.name');
```

davi5e commented:
This is all a simplification, though.

My real scenario is more like the following:
```
https://api.example.com/service-1/vX/...
https://api.example.com/service-2/vX/...
```

In which services are used as prefixes in every route:
```php
    Route::group([
        'as'        => 'service.vX.',
        'middleware' => 'api.v:X,service',
        'namespace' => 'Service\vX\Model',
        'prefix'    => 'service/vX',
    ], static function() {
        // Resulting name: service.vX.model.something
        // Resulting namespace: 'App\Http\Resources\Service\vX\Model (this one uses what's defined in RouteServiceProvider as well)
        // Resulting URL: <domain>/service/vX/model
        Route::get('/model', 'ModelController@something')
            ->name('model.something');
    }
```

And this file would be `app/Http/Resources/Service/vX/Model/ModelController.php`. Keep in mind the actual use case has names that make more sense than Model :)

juampi92 commented:
@davi5e, based on what you are saying, the problem is the following:

```
The package's helper api_route() doesn't work with multiple apis (or domains or services)
```

If that's what you are saying, that's not correct. This package is supposed to attend one api (or domain or service) at a time.

So as you said on your example:
This endpoint `https://api.example.com/service-1/vX/` has a middleware `api.v:X,service-1`.

That means that if you try to do `api_route('service-1.login')` **inside that call**, it will output `route('service-1.vX.login')`.

The problem is, if you are accessing that route from another service, imagine service-2, it won't work.

Calling `api_route('service-2.login')` **inside a route that has the middleware `api.v:X,service-1`**, it will output `route('service-2.login')`.

This is because you **don't know the current version of service-2**.
The best fix is to use on your resources a static versioning call: `route('service-2.vY.login')`. This way the package won't have to figure out the correct version of service-2 for the service-1 vX.

davi5e commented:
Indeed!

IMHO that whole comment should be in the doc...

The problem was the Middleware scope :)

juampi92 commented:
Will update! Thanks man!
